### ISSUE TITLE
Support TeaCache in Memo

### ISSUE THREAD

**Issue by wallkop** [ali-vilab/TeaCache#22]
Memo(https://github.com/memoavatar/memo) is an excellent digital human project with impressive results.
However, its performance is quite slow. Can TeaCache support Memo?

**Comment by LiewFeng** [comment]
You can try it following [issue 20](https://github.com/ali-vilab/TeaCache/issues/20). Welcome for a PR to support it.

**Comment by wallkop** [comment]
> You can try it following [issue 20](https://github.com/ali-vilab/TeaCache/issues/20). Welcome for a PR to support it.

I've tried writing a teacache-accelerated wrapper for memo. It performs fairly well, achieving about a 2x speedup with a quality loss within an acceptable range at rel_l1_thresh=0.15.

I'm sharing the code here and would appreciate it if you could review it. If I were to submit a PR, what additional information would I need to provide? I'm willing to offer further assistance.

```
import torch
from memo.pipelines.video_pipeline import VideoPipeline
from memo.models.unet_3d import UNet3DConditionOutput


def teacache_forward(
        self,
        sample: torch.FloatTensor,
        ref_features: dict,
        timestep: torch.Tensor,
        encoder_hidden_states: torch.Tensor,
        audio_embedding: torch.Tensor = None,
        audio_emotion: torch.Tensor = None,
        class_labels: torch.Tensor = None,
        mask_cond_fea: torch.Tensor = None,
        attention_mask: torch.Tensor = None,
        down_block_additional_residuals: tuple = None,
        mid_block_additional_residual: torch.Tensor = None,
        uc_mask: torch.Tensor = None,
        return_dict: bool = True,
        is_new_audio: bool = True,
        update_past_memory: bool = False,
):
    """
    TeaCache-accelerated forward for UNet3DConditionModel.

    This function retains the original signature to remain compatible with the pipeline's calling logic.
    It implements a simplified TeaCache strategy:

    1) Right after `conv_in`, it extracts an intermediate activation (called `modulated_inp`) and compares
       it with the previous activation to decide whether the current step should be skipped.
    2) If the difference from the previous step is small enough and we have a cached final output,
       it returns that cached output directly (i.e., skipping the entire network computation for this step).
    3) Otherwise, it performs the full forward pass (down blocks -> mid block -> up blocks -> final conv_out)
       and updates the cached output for future potential skips.
    """

    # --- Pre-processing (mirrors the original forward) ---
    default_overall_up_factor = 2 ** self.num_upsamplers
    forward_upsample_size = False
    upsample_size = None

    # Determine if we should forcibly set an upsample size
    # when the spatial dimensions aren't multiples of the upsampling factor
    if any(s % default_overall_up_factor != 0 for s in sample.shape[-2:]):
        forward_upsample_size = True

    # Prepare the attention mask (if any)
    if attention_mask is not None:
        attention_mask = (1 - attention_mask.to(sample.dtype)) * -10000.0
        attention_mask = attention_mask.unsqueeze(1)

    # Optionally center the input sample based on model config
    if self.config.center_input_sample:
        sample = 2 * sample - 1.0

    # Time embeddings
    if not torch.is_tensor(timestep):
        timestep = torch.tensor([timestep], dtype=sample.dtype, device=sample.device)
    elif len(timestep.shape) == 0:
        timestep = timestep[None].to(sample.device)

    # Expand the timestep to match batch size
    timestep = timestep.expand(sample.shape[0])
    t_emb = self.time_proj(timestep).to(dtype=self.dtype)
    emb = self.time_embedding(t_emb)

    # Class embedding (if applicable)
    if self.class_embedding is not None:
        if class_labels is None:
            raise ValueError("class_labels must be provided when num_class_embeds > 0.")
        if self.config.class_embed_type == "timestep":
            class_labels = self.time_proj(class_labels)
        class_emb = self.class_embedding(class_labels).to(dtype=self.dtype)
        emb = emb + class_emb

    # Initial convolution
    sample = self.conv_in(sample)
    if mask_cond_fea is not None:
        sample = sample + mask_cond_fea

    # --- TeaCache check: decide whether to skip this step ---
    if getattr(self, "enable_teacache", False):
        modulated_inp = sample.clone()

        # Force compute on step=0 or step=(num_steps-1), or else compare to the previous input
        if self.cnt == 0 or self.cnt == self.num_steps - 1:
            should_calc = True
            self.accumulated_rel_l1_distance = 0
        else:
            if self.previous_modulated_input is not None:
                rel_l1 = (
                        (modulated_inp - self.previous_modulated_input).abs().mean()
                        / (self.previous_modulated_input.abs().mean() + 1e-8)
                ).item()
            else:
                # If we have no previous_modulated_input, treat it as large difference
                rel_l1 = 999.0

            self.accumulated_rel_l1_distance += rel_l1

            if self.accumulated_rel_l1_distance < self.rel_l1_thresh:
                should_calc = False
            else:
                should_calc = True
                self.accumulated_rel_l1_distance = 0

        # Update the intermediate activation
        self.previous_modulated_input = modulated_inp

        # Update the step counter and reset if we reach num_steps
        self.cnt += 1
        if self.cnt == self.num_steps:
            self.cnt = 0

        # If changes are small enough and there's a cached final output, skip
        if (not should_calc) and (self.previous_final_output is not None):
            if not return_dict:
                return (self.previous_final_output,)
            else:
                return UNet3DConditionOutput(sample=self.previous_final_output)
        else:
            # Need to do a full forward pass
            do_calc = True
    else:
        do_calc = True

    # --- Normal UNet3D logic: down blocks, mid block, up blocks ---
    # down blocks
    down_block_res_samples = (sample,)
    for i, downsample_block in enumerate(self.down_blocks):
        if hasattr(downsample_block, "has_cross_attention") and downsample_block.has_cross_attention:
            sample, res_samples, audio_embedding = downsample_block(
                hidden_states=sample,
                ref_feature_list=ref_features["down"][i],
                temb=emb,
                encoder_hidden_states=encoder_hidden_states,
                attention_mask=attention_mask,
                audio_embedding=audio_embedding,
                emotion=audio_emotion,
                uc_mask=uc_mask,
                is_new_audio=is_new_audio,
                update_past_memory=update_past_memory,
            )
        else:
            sample, res_samples = downsample_block(
                hidden_states=sample,
                ref_feature_list=ref_features["down"][i],
                temb=emb,
                encoder_hidden_states=encoder_hidden_states,
                is_new_audio=is_new_audio,
                update_past_memory=update_past_memory,
            )
        down_block_res_samples += res_samples

    # Add any optional down-block residuals
    if down_block_additional_residuals is not None:
        new_down_block_res_samples = ()
        for db_res_sample, db_add_res in zip(down_block_res_samples, down_block_additional_residuals):
            db_res_sample = db_res_sample + db_add_res
            new_down_block_res_samples += (db_res_sample,)
        down_block_res_samples = new_down_block_res_samples

    # mid block
    sample, audio_embedding = self.mid_block(
        sample,
        ref_feature_list=ref_features["mid"][0],
        temb=emb,
        encoder_hidden_states=encoder_hidden_states,
        attention_mask=attention_mask,
        audio_embedding=audio_embedding,
        emotion=audio_emotion,
        uc_mask=uc_mask,
        is_new_audio=is_new_audio,
        update_past_memory=update_past_memory,
    )
    if mid_block_additional_residual is not None:
        sample = sample + mid_block_additional_residual

    # up blocks
    for i, upsample_block in enumerate(self.up_blocks):
        is_final_block = i == len(self.up_blocks) - 1
        res_samples = down_block_res_samples[-len(upsample_block.resnets):]
        down_block_res_samples = down_block_res_samples[: -len(upsample_block.resnets)]

        if not is_final_block and forward_upsample_size:
            upsample_size = down_block_res_samples[-1].shape[2:]

        if hasattr(upsample_block, "has_cross_attention") and upsample_block.has_cross_attention:
            sample, audio_embedding = upsample_block(
                hidden_states=sample,
                ref_feature_list=ref_features["up"][i],
                temb=emb,
                res_hidden_states_tuple=res_samples,
                encoder_hidden_states=encoder_hidden_states,
                upsample_size=upsample_size,
                attention_mask=attention_mask,
                audio_embedding=audio_embedding,
                emotion=audio_emotion,
                uc_mask=uc_mask,
                is_new_audio=is_new_audio,
                update_past_memory=update_past_memory,
            )
        else:
            sample = upsample_block(
                hidden_states=sample,
                ref_feature_list=ref_features["up"][i],
                temb=emb,
                res_hidden_states_tuple=res_samples,
                upsample_size=upsample_size,
                encoder_hidden_states=encoder_hidden_states,
                is_new_audio=is_new_audio,
                update_past_memory=update_past_memory,
            )

    # Final post-processing
    sample = self.conv_norm_out(sample)
    sample = self.conv_act(sample)
    sample = self.conv_out(sample)

    output = sample  # shape [B, out_channels, F, H, W]

    # If we actually computed this step (instead of skipping), cache for next time
    if getattr(self, "enable_teacache", False) and do_calc:
        self.previous_final_output = output.detach()

    if not return_dict:
        return (sample,)

    return UNet3DConditionOutput(sample=sample)


def teacache_wrap(pipeline: VideoPipeline, num_steps: int, rel_l1_thresh=0.15):
    """
    Patches the pipeline's diffusion_net (UNet3DConditionModel) by replacing its forward method
    with teacache_forward, enabling a lightweight TeaCache mechanism.

    :param pipeline: The VideoPipeline instance containing diffusion_net as its 3D UNet.
    :param num_steps: The number of diffusion steps, used by the TeaCache logic.
    :param rel_l1_thresh: The relative L1 threshold below which we consider the change to be insignificant.
    """
    pipeline.diffusion_net.__class__.enable_teacache = True
    pipeline.diffusion_net.__class__.cnt = 0
    pipeline.diffusion_net.__class__.num_steps = num_steps
    pipeline.diffusion_net.__class__.rel_l1_thresh = rel_l1_thresh
    pipeline.diffusion_net.__class__.accumulated_rel_l1_distance = 0
    pipeline.diffusion_net.__class__.previous_modulated_input = None
    pipeline.diffusion_net.__class__.previous_final_output = None

    pipeline.diffusion_net.__class__.forward = teacache_forward

    print("[TeaCache] Successfully replaced the UNet3DConditionModel forward with teacache_forward!")

```

**Comment by LiewFeng** [comment]
You can refer to [this](https://github.com/ali-vilab/TeaCache/tree/main/TeaCache4HunyuanVideo). A new directory with one python file and readme. You can launch a PR. I will review the PR.
